prow/plugins/approve_configtree_test.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugins
+
+import (
+	"testing"
+
+	"sigs.k8s.io/yaml"
+)
+
+func TestApproveConfigTree(t *testing.T) {
+	var cases = []struct {
+		name        string
+		config      []byte
+		expectError bool
+	}{
+		{
+			name: "approve no orgs",
+			config: []byte(`
+config:
+  issue_required: false
+  require_self_approval: true
+  lgtm_acts_as_approve: false
+  ignore_review_state: true
+`),
+			expectError: true,
+		},
+		{
+			name: "approve no default",
+			config: []byte(`
+orgs:
+  bazelbuild:
+    config:
+      ignore_review_state: false
+  kubernetes:
+    config:
+      lgtm_acts_as_approve: true
+    repos:
+      kops:
+        config:
+          lgtm_acts_as_approve: false
+      kubernetes:
+        config:
+          require_self_approval: true
+`),
+		},
+		{
+			name: "approve full",
+			config: []byte(`
+config:
+  issue_required: false
+  require_self_approval: false
+  lgtm_acts_as_approve: false
+  ignore_review_state: true
+orgs:
+  bazelbuild:
+    config:
+      ignore_review_state: false
+  kubernetes:
+    config:
+      lgtm_acts_as_approve: true
+    repos:
+      kops:
+        config:
+          lgtm_acts_as_approve: false
+      kubernetes:
+        config:
+          require_self_approval: true
+        branches:
+          master:
+            require_self_approval: false
+`),
+		},
+	}
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			var tree ApproveConfigTree
```

Comment:
Where is this type defined?

Reply:
Well, this is the dynamically created file out of the template... look at BUILD.bazel

Reply:
How does this work for devs iterating on things? Do I need to run `bazel` to have something that compiles and/or follow references?

Reply:
Well, the file `approve_configtree.go` only exists during the bazel build... so without it it's awkward to work.
You could run bazel in debug sandbox, and copy the generated file from the temporary workspace.

Reply:
Can we commit the file? What do we gain by making it ephemeral?

Reply:
This lib was made to not commit the generated files... Not sure how the team at Gvisor handle it.

Otherwise we could try another one in a different branch.

Reply:
@cjwagner thoughts? I'm not the biggest fan :|

Reply:
Well, it brings some advantages:
- no need to have a complex workflow to generate, update and commit files (like the protobuf in k/k)
- the tool is completely integrated into bazel
- you can still write unit tests (this is what I did)
- it's Google (not sure it's a plus or a con)

And some disadvantages too:
- only works with bazel
- you need bazel to debug

Maybe we could ask some Gvisor devs on how they work with that tool, and if they advise using it in our case?

Reply:
I'm also not the biggest fan, but given that the code won't even build without using bazel to generate the concrete types I'm not sure that this adds any additional down sides.

I was thinking we could test without generation if we write unit tests for `configtree.go` by specifying some fields for `T` and adding a `configtree_test.go` that tests the merging behavior. If we also unit tested `Approve.Apply()` we would then have complete unit test coverage, we'd just be missing the integration test of the generated concrete type which might be acceptable.
However since the plugin config struct can't compile without the generated concrete type I think this is a moot point.

If we can't come up with a better way to do the testing I'm fine with trying this out and seeing if it causes problems.